libcextract/ElfCXX.cpp

```diff
@@ -91,12 +96,97 @@ ElfObject::ElfObject(const char *path)
     throw std::runtime_error("ELF file not found: " + parser_path);
   }
 
-  /* Create libelf object and store it in the class variable.  */
-  ElfObj = elf_begin(ElfFd, ELF_C_READ, nullptr);
-  if (ElfObj == nullptr) {
+  unsigned char buf[4];
+
+  read(ElfFd, buf, 4);
+  /* reset file pointer */
+  lseek(ElfFd, 0, SEEK_SET);
+
+  /* Check if the file is an ELF */
+  if (buf[0] == 0x7f && buf[1] == 0x45 && buf[2] == 0x4c && buf[3] == 0x46) {
+    /* Create libelf object and store it in the class variable.  */
+    ElfObj = elf_begin(ElfFd, ELF_C_READ, nullptr);
+    if (ElfObj == nullptr) {
+      close(ElfFd);
+      throw std::runtime_error("libelf error on file " + parser_path + ": "
+                                   + std::string(elf_errmsg(elf_errno())));
+    }
+  /* gzip magic number (zlib) */
+  } else if (buf[0] == 0x1f && buf[1] == 0x8b) {
+    struct stat st;
+
+    gzFile gzfile = gzopen(parser_path.c_str(), "r");
+    if (!gzfile)
+      throw std::runtime_error("could not gzopen " + parser_path + "\n");
+
+    int ret = fstat(ElfFd, &st);
+    if (ret)
+      throw std::runtime_error("stat on " + parser_path + " failed\n");
+
+    /* It seems that the extracted file can be 5 times bigger than the
+     * compressed time */
+    unsigned long deflate_len = st.st_size * 5;
```

Comment:
And what evidence do you have to support this claim? That is not how you should read a compressed archive.

Compressed archives either have in its header the size of the original file it will deflate (which is dangerous!), or it needs to be deflated in a stream.

Looking at the `gzread` documentation, it seems you can pass the buffer size as a parameter. From there you make your guess and try to read the file. If it returns that it did not manage to read the entire file, then you `realloc` the buffer at the larger size and try again in a loop from where it stopped.

Here is an example from libpulp's code:

https://github.com/SUSE/libpulp/blob/c1704b4d99a762a27b5e12c5763cbc3986de8e1e/tools/ptrace.c#L375